---
title: "RL RNA-seq"
output: html_notebook
---

#install packages and load libraries
```{r eval=FALSE}
#source("https://bioconductor.org/biocLite.R")
#biocLite()
#biocLite("edgeR")
#biocLite("limma")
install.packages("statmod")
library(statmod)
library(edgeR)
library(limma)
```

#load files
```{r}
files <- c("S01_N2_L2_15_mk", "S02_N2_L2_15_nd", "S03_N2_L2_15_np", "S04_N2_L2_20_mk",
           "S05_N2_L2_20_nd", "S06_N2_L2_20_np", "S07_N2_L4_15_mk", "S08_N2_L4_15_nd",
           "S09_N2_L4_15_np", "S10_N2_L4_20_mk", "S11_N2_L4_20_nd", "S12_N2_L4_20_np",
           "S13_N2_L4_25_mk", "S14_N2_L4_25_nd", "S15_N2_L4_25_np")
files <- paste(files, "_htseq-count.txt", sep = "")
table <- read.table(files[1], col.names = c("WBGeneID", "count"))
head(table)
dim(table)
```

#Create DGElist with count data. Follow up with data check.
note: `readDGE()` is used for multiple files like here. If all data is contained in a single matrix (or other) file then use the `DGEList()`
```{r}
x <- readDGE(files)
class(x)
dim(x)
head(x$samples)
head(x$counts)
```
#We will also load matrix file for N. parisii + N. ausebeli + N2 untreated RNA-seq data.
These data were obtained in biological quadlicate (tetracate?) I am going to load them into a seperate DGEList object since, at the moment, I am not sure if it is proper to analyze the data sets simultaneously.
Additionally, since the transcripts in this file are labelled with common names or locus IDs, it is difficult to match with the other data labelled with WBGene IDs. This is potentially something that can be solved during mapping.
```{r}
files2 <- read.csv("nematocida.counts.matrix", stringsAsFactors = F, check.names = F)
count_matrix <- as.matrix(files2[,-1])
rownames(count_matrix) <- files2[,1]
head(count_matrix)
x2 <- DGEList(counts = count_matrix)
class(x2)
head(x2$counts)
x2$samples
```

#Organize sample data 
```{r}
#temp <- as.factor(c(rep("x15C",3),rep("x20C",3),rep("x15C",3),rep("x20C",3),rep("x25C",3)))
temp <- as.factor(rep(c("x15C", "x20C", "x15C", "x20C", "x25C"), c(3,3,3,3,3)))
temp <- relevel(temp, ref = "x20C")
x$samples$temp <- temp
stage <- as.factor(rep(c("L2","L4"),c(6,9)))
stage <- relevel(stage, ref="L4")
x$samples$stage <- stage
treatment <- as.factor(rep(c("mock","parisii","displodere"),5))
treatment <- relevel(treatment, ref="mock")
x$samples$group <- treatment
head(x$samples)

#########

treatment2 <- as.factor(c(rep(c("mock", "parisii", "ausubeli"),3),"parisii", "mock", "ausubeli"))
x2$samples$group <- treatment2
lane <- as.factor(c(rep("L004",9),rep("L007",3)))
x2$samples$lane <- lane
head(x2$samples)
```
#Data pre-processing - converting data from raw counts.
```{r}
cpm <- cpm(x)
lcpm <- cpm(x, log=T)

#############

cpm2 <- cpm(x2)
lcpm2 <- cpm(x2, log=T)
```

#Identify genes that are lowly expressed. 
This expression will show genes that have ZERO reads across all samples. In our case its ~60% of all genes.
```{r}
table(rowSums(x$counts==0)==15) 
#Values returned: FALSE=20942 TRUE=25801

table(rowSums(x2$counts==0)==16)
#201535=False
```

#Remove genes that are lowly expressed. 
We set our threshold to cpm>1 in at least one sample due to our limitation of one replicate. Remember that cpm is dependent on library size, so the raw read count threshold will be unique for each sample. Following this action, we will reduce our total genes down significantly, and the library sizes will change as well.
```{r}
keep.exprs <- rowSums(cpm>1)>=1
x <- x[keep.exprs,, keep.lib.sizes=F]
dim(x)

##################

keep.exprs <- rowSums(cpm2>1)>=4
x2 <- x2[keep.exprs,, keep.lib.sizes=F]
```
#Density plot showing pre-filtered vs filtered data.
```{r}
library(RColorBrewer)
nsamples <- ncol(x)
col <- brewer.pal(nsamples, "Set3")
par(mfrow=c(1,2))
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Raw data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
#legend("topright", row.names(x$samples), text.col=col, bty="n")
lcpm <- cpm(x,log=T)
plot(density(lcpm[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Filtered data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
#legend("topright", rownames(x$samples), text.col = col, bty = "n")

##########################

library(RColorBrewer)
nsamples <- ncol(x2)
par(mfrow=c(1,2))
plot(density(lcpm2[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Raw data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm2[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
#legend("topright", row.names(x2$samples), text.col=col, bty="n")
lcpm2 <- cpm(x2,log=T)
plot(density(lcpm2[,1]), col=col[1], lwd=2, ylim=c(0,0.21), las=2, main="", xlab="")
title(main="Filtered data", xlab="log-cpm")
abline(v=0, lty=3)
for (i in 2:nsamples) {
  den <- density(lcpm2[,i])
  lines(den$x, den$y, col=col[i], lwd=2)
}
#legend("topright", rownames(x2$samples), text.col = col, bty = "n")
```

#Normalization
This normalization weights the samples based on library size. "TMM normalization is performed to eliminate composition bias between libraries."
```{r}
x <- calcNormFactors(x, method = "TMM")
x$samples$norm.factors

######################
x2 <- calcNormFactors(x2, method="TMM")
x2$samples$norm.factors
```

#Unsupervised clustering
```{r}
col.1 <- treatment
col.2 <- stage
col.3 <- temp

samplesnames <- rownames(x$samples)
samplesnames <- substring(samplesnames,8,15)

levels(col.1) <- brewer.pal(nlevels(col.1), "Set1")
col.1 <- as.character(col.1)

levels(col.2) <- brewer.pal(nlevels(col.2), "Set2")
col.2 <- as.character(col.2)

levels(col.3) <- brewer.pal(nlevels(col.3), "Dark2")
col.3 <- as.character(col.3)

lcpm <- cpm(x,log=T)
par(mfrow=c(1,2))

plotMDS(lcpm, labels=samplesnames, col=col.1, cex=0.66)
title(main = "x1. Treatment")
plotMDS(lcpm, labels=samplesnames, col=col.1, cex=0.66, dim=c(3,4))
title(main = "x1. Treatment")

plotMDS(lcpm, labels=samplesnames, col=col.2, cex=0.66)
title(main = "x1. Stage")
plotMDS(lcpm, labels=samplesnames, col=col.2, cex=0.66, dim=c(3,4))
title(main = "x1. Stage")

plotMDS(lcpm, labels=samplesnames, col=col.3, cex=0.66)
title(main = "x1. Temp")
plotMDS(lcpm, labels=samplesnames, col=col.3, cex=0.66, dim=c(3,4))
title(main = "x1. Temp")

########################

col.1 <- x2$samples$group
col.2 <- x2$samples$lane

levels(col.1) <- brewer.pal(nlevels(col.1), "Set1")
col.1 <- as.character(col.1)

levels(col.2) <- brewer.pal(nlevels(col.2), "Set2")
col.2 <- as.character(col.2)

plotMDS(lcpm2, labels = x2$samples$group, col=col.1, cex=0.66)
title(main="x2. Treatment")

plotMDS(lcpm2, labels = x2$samples$group, col=col.1, cex=0.66, dim=c(3,4))
title(main="x2. Treatment")

plotMDS(lcpm2, labels = x2$samples$lane, col=col.2, cex=0.66)
title(main = "x2. Lane")

plotMDS(lcpm2, labels = x2$samples$lane, col=col.2, cex=0.66, dim=c(3,4))
title(main = "x2. Lane")
```

```{r}
design <- model.matrix(~0+treatment+stage+temp)
colnames(design) <- gsub("treatment","",colnames(design))
colnames(design) <- gsub("temp","",colnames(design))
colnames(design) <- gsub("stage","",colnames(design))
design

##################

design2 <- model.matrix(~0+treatment2+lane)
colnames(design2) <- gsub("treatment2","",colnames(design2))
design2
```

#Estimate dispersion
This is a critical juncture for analyzing RNA-seq with no biological replicates. Disperion is essentially the estimated biological variation. Without replicates, dispersion cannot be estimated and must be manually entered. For more information, read the edgeR User Guide, section 2.11. 
For our purposes, since we have the other data with replicates, I will calculate dispersion using those data and use that value. An alternative option, is to look at the MDS plots above, find closely clustered samples, and pretend those are replicates in order to estimate dispersion.
```{r}
x2 <- estimateDisp(x2, design2, robust = T)
x$common.dispersion <- x2$common.dispersion
plotBCV(x2)
#bcv
```

```{r}
fit <- glmFit(x, design)
################
fit2 <- glmFit(x2, design2)

head(fit$coefficients)
head(fit2$coefficients)
```
#Make contrasts
```{r}
contrast.matrix <- makeContrasts(ndvnp = displodere - parisii, levels = design)
contrast.matrix2 <- makeContrasts(Navmk = ausubeli - mock, 
                                  Npvmk = parisii - mock, 
                                  NavNp = ausubeli - parisii, 
                                  levels = design2)

df <- glmLRT(fit, contrast = contrast.matrix)

topTags(df, n=30, sort.by = "logFC")
summary(decideTests(df))
plotMD(df)
```
#This method is adapted from the edgeR User Guide 3.5. 
```{r}
design <- model.matrix(~treatment+treatment:stage+treatment:temp)
colnames(design)

x <-estimateDisp(x,design,robust=T)
fit <- glmQLFit(x,design)
```
#Genes differentially expressed in "mock" treated animals L2 compared to L4
```{r}
qlf <- glmQLFTest(fit, coef = "treatmentmock:stageL2")
summary(decideTests(qlf))
topTags(qlf, n=50, sort.by = "logFC")
```
The top hits by absolute logFC is WBGene00003435, WBGene00022754. WBGene00044058 has a large positive logFC. Use these three genes for a sanity check, make sure the raw count data performs as expected. 
```{r}
x$counts["WBGene00003435",]
x$counts["WBGene00022754",]
x$counts["WBGene00044058",]
```
Genes that respond to disp. at ANY temp or stage compared to mock
```{r}
qlf <- glmQLFTest(fit, coef = c(5,8,11))
topTags(qlf, n=50, sort.by = "PValue")
summary(decideTests(qlf))
```
sanity check again
```{r}
x$counts["WBGene00016541",]
x$counts["WBGene00014158",]
x$counts["WBGene00010516",]
```
Compare disp at 15 to disp at 20
```{r}
qlf <- glmQLFTest(fit, coef = 8)
topTags(qlf, n=50, sort.by = "logFC")
summary(decideTests(qlf))
```
#Compare dispoldere to parisii, note: this is across all stages and temps
```{r}
qlf <- glmQLFTest(fit, contrast = c(0,1,-1,0,0,0,0,0,0,0,0,0))
topTags(qlf, n=50, sort.by = "logFC")
summary(decideTests(qlf))
```
#Compare L2 20C disp vs parisii
```{r eval=F}
qlf <- glmQLFTest(fit,coef = c)
topTags(qlf, sort.by = "logFC")
summary(decideTests(qlf))
```

```{r}
qlf <- glmQLFTest(fit,coef = c(5,8,11))
topTags(qlf, sort.by = "logFC")
summary(decideTests(qlf))
```
#Exact Test
compare two samples without a linear model
```{r}
#First, redesign the group parameter to make each sample an individual group
x$samples$group <- 1:15
x$samples

#Set bcv
bcv <- 0.2

# L2 20C disp - N2
et1 <- exactTest(x, pair = 4:5, dispersion = bcv^2)
topTags(et, n=Inf, sort.by = "logFC")
df1_name <- "L2.20C"

#L4 25C disp - N2
et2 <- exactTest(x, pair=13:14, dispersion = bcv^2)
topTags(et, n=Inf, sort.by = "logFC")
df2_name <- "L4.25C"

#L4 20C disp - N2
et3 <- exactTest(x, pair=10:11, dispersion=bcv^2)
topTags(et, n=Inf, sort.by = "logFC")
df3_name <- "L4.20C"

#Generate data frame and print to csv
logfc.df <- data.frame(et1$table$logFC,et2$table$logFC,et3$table$logFC, 
                       row.names = rownames(et1$table), stringsAsFactors = F)
names(logfc.df) <- c(df1_name,df2_name,df3_name)
head(logfc.df)
write.csv(logfc.df, file = "2018_07_27_Disp_N2_logFC.csv")
```


Alternative experiment design
```{r}
group <- factor(make.names(treatment:stage:temp))
design <- model.matrix(~0+group)
colnames(design)

contrast.matrix <- makeContrasts(disp.L4.20 = groupdisplodere.L4.x20C - groupmock.L4.x20C,
                                 disp.L2.20 = groupdisplodere.L2.x20C - groupmock.L2.x20C,
                                 pari.L4.20 = groupparisii.L4.x20C - groupmock.L4.x20C,
                                 pari.L2.20 = groupparisii.L2.x20C - groupmock.L2.x20C,
                                 dispvpari.L4.25 = groupdisplodere.L4.x25C - groupparisii.L4.x25C,
                                 levels = design
)

bcv <- 0.1

fit <- glmFit(x,design,dispersion = bcv^2)
lrt <- glmLRT(fit, contrast = contrast.matrix[,5])
topTags(lrt)
summary(decideTests(lrt))
```

#Continue linear modeling and comparison of the triplicate data
```{r}
v <- voom(x2, design2, plot=T)
v
vfit <- lmFit(v, design2)
vfit <- contrasts.fit(vfit, contrasts = contrast.matrix2)
efit <- eBayes(vfit)
plotSA(efit)
summary(decideTests(efit))
```
```{r}
tops <- topTable(efit, coef = 2, n=Inf, sort.by = "logFC", p.value = 0.05)
tops
#write.csv(tops, "2018_07_27_parisii_topgenes.csv")
```

